Ardupilog.m

```diff
@@ -293,6 +312,65 @@ function delete(obj)
             end
         end
         
+        function [] = findBootTimeUTC(obj)
+        % From the GPS time, stored in GWk and GMS, calculate what UTC
+        % (Coordinated Universal time) was when the Ardupilot microcontroller
+        % booted. (TimeUS = AP_HAL::millis() = 0)
+            
+        % HGM: It's possible the accuracy of this can be improved. I'll put the
+        % details of my idea here in the comments, and we can move to a GitHub
+        % issue or whatever as appropriate.
+        %
+        % When a GPS receives data, containing absolute time info (logged in GWk and
+        % GMS) it is timestamped by Ardupilot in microseconds-since-boot.  The
+        % problem is, that timestamp is stored in the GPA log message, while the
+        % GWk/GMS is stored in the GPS log message. The delay between logging the two
+        % (GPS and GPA messages) is probably small, but I don't know if there's any
+        % way to determine what came from the single original data receipt.
+        %
+        % We could ask this be changed in Ardupilot, or we might implement
+        % something to figure it out from the log... for now, I'm neglecting it,
+        % and assuming the GPS message was RECEIVED at it's TimeUS. (Note: the
+        % truth is it was LOGGED at this time, not received)
+            if isprop(obj, 'GPS') && ~isempty(obj.GPS.TimeUS)
+                % Get the time data from the log
+                recv_timeUS = obj.GPS.TimeUS(1);
+                recv_GWk = obj.GPS.GWk(1);
+                recv_GMS = obj.GPS.GMS(1);
+                % Calculate the gps-time datenum
+                gps_zero_datenum = datenum('1980-01-06 00:00:00.000','yyyy-mm-dd HH:MM:SS.FFF');
+                days_since_gps_zero = recv_GWk*7 + recv_GMS/1e3/60/60/24;
+                recv_gps_datenum = gps_zero_datenum + days_since_gps_zero;
+                % Adjust for leap seconds (disagreement between GPS and UTC)
+                leap_second_table = datenum(...
+                    ['Jul 01 1981'
+                     'Jul 01 1982'
+                     'Jul 01 1983'
+                     'Jul 01 1985'
+                     'Jan 01 1988'
+                     'Jan 01 1990'
+                     'Jan 01 1991'
+                     'Jul 01 1992'
+                     'Jul 01 1993'
+                     'Jul 01 1994'
+                     'Jan 01 1996'
+                     'Jul 01 1997'
+                     'Jan 01 1999'
+                     'Jan 01 2006'
+                     'Jan 01 2009'
+                     'Jul 01 2012'
+                     'Jul 01 2015'], 'mmm dd yyyy');
```

Comment:
Can you cite this from somewhere? Wheren't sometimes leap seconds removed, instead of added?